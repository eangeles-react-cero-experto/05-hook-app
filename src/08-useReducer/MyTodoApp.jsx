import { useReducer } from "react";
import { todoListReducer } from "./todoListReducer.js";

const initialState = [
  {
    id: crypto.randomUUID(),
    description: "Collect the soul stone",
    isDone: false,
  },
  {
    id: crypto.randomUUID(),
    description: "Collect the power stone",
    isDone: false,
  },
];

export const MyTodoApp = () => {
  const [state, dispatch] = useReducer(todoListReducer, initialState);

  return (
    <>
      <h1>MyTodoApp (10)</h1>
      <hr />

      <div className="row">
        <div className="col-6">
          {/* TodoList */}
          <ul className="list-group">
            {/* TodoItem */}
            <li className="list-group-item d-flex justify-content-between align-items-center">
              <span>Item 1</span>
              <button className="btn btn-sm btn-outline-danger">Remove</button>
            </li>
            {/* TodoItem */}
            <li className="list-group-item">Item 2</li>
            <li className="list-group-item">Item 3</li>
          </ul>
          {/* TodoList */}
        </div>
        <div className="col-6">
          <h4>Add Todo</h4>
          <hr />
          {/* FromAddTodo onNewTodoItemAdded */}
          <form>
            <input
              type="text"
              placeholder="Description here"
              className="form-control"
            />

            <button type="submit" className="btn btn-outline-primary mt-3">
              Add Item
            </button>
          </form>
          {/* FormAddTodo */}
        </div>
      </div>
    </>
  );
};
