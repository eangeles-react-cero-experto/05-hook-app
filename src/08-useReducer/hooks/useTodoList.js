import { useEffect, useReducer } from "react";
import { todoListReducer } from "../todoListReducer.js";

const initializer = () => {
  return JSON.parse(localStorage.getItem("todoList") || "[]");
};

export const useTodoList = (initialValue = []) => {
  const [todoList, dispatch] = useReducer(
    todoListReducer,
    initialValue,
    initializer,
  );

  useEffect(() => {
    localStorage.setItem("todoList", JSON.stringify(todoList));
  }, [todoList]);

  const addNewTodoItem = (todoItem) => {
    const action = {
      type: "[TODO] Add todo item",
      payload: todoItem,
    };
    dispatch(action);
  };

  const deleteTodoItem = (id) => {
    const action = {
      type: "[TODO] Delete todo item",
      payload: id,
    };
    dispatch(action);
  };

  const toggleTodoItem = (id) => {
    const action = {
      type: "[TODO] Toggle todo item",
      payload: id,
    };
    dispatch(action);
  };

  const todoItemsAmount = todoList.length;
  const completedTodoItemsAmount = todoList.filter((x) => x.isDone).length;
  const pendingTodoItemsAmount = todoList.filter((x) => !x.isDone).length;

  return {
    todoList,
    todoItemsAmount,
    completedTodoItemsAmount,
    pendingTodoItemsAmount,
    addNewTodoItem,
    deleteTodoItem,
    toggleTodoItem,
  };
};
