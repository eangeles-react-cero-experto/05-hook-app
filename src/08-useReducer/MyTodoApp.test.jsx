import { useTodoList } from "./hooks/useTodoList.js";
import * as crypto from "crypto";
import { render, screen } from "@testing-library/react";
import { MyTodoApp } from "./MyTodoApp.jsx";

jest.mock("./hooks/useTodoList.js");

describe("Test MyTodoApp", () => {
  const todoList = [
    { id: crypto.randomUUID(), description: "Walk the dog", isDone: false },
    { id: crypto.randomUUID(), description: "Make dinner", isDone: true },
  ];
  useTodoList.mockReturnValue({
    todoList,
    todoItemsAmount: 2,
    completedTodoItemsAmount: 1,
    addNewTodoItem: jest.fn(),
    deleteTodoItem: jest.fn(),
    toggleTodoItem: jest.fn(),
  });

  it("should match with snapshot", () => {
    const { container } = render(<MyTodoApp />);
    expect(container).toMatchSnapshot();
  });

  it("should render render todo items", () => {
    render(<MyTodoApp />);
    expect(screen.getByText(todoList[0].description)).toBeTruthy();
    expect(screen.getByText(todoList[1].description)).toBeTruthy();
  });
});
