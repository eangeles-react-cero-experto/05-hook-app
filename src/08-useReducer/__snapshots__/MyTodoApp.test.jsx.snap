// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MyTodoApp should match with snapshot 1`] = `
<div>
  <h1>
    MyTodoApp Â 
    <small>
      (
      1
       / 
      2
      )
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-6"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between align-items-center"
        >
          <span
            class="w-100 "
          >
            Walk the dog
          </span>
          <button
            class="btn btn-sm btn-outline-danger"
          >
            Remove
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between align-items-center"
        >
          <span
            class="w-100 text-decoration-line-through"
          >
            Make dinner
          </span>
          <button
            class="btn btn-sm btn-outline-danger"
          >
            Remove
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-6"
    >
      <h4>
        Add Todo
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          placeholder="Description here"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
          type="submit"
        >
          Add Item
        </button>
      </form>
    </div>
  </div>
</div>
`;
